### PR TITLE
Weekly openmeteo extract.py: Alex

### LINKED ISSUE #23: Create test file for weekly meteo API extract script
As a developer, I want to test the extract.py file to ensure that the data is being extracted from APIs correctly and robustly, and that the constituent functions deal with appropriate edge cases and erroneous input effectively, so that the user experience is as intended.

### PR DESCRIPTION
**Pull request by AlexOBarnes**
This closes #23 #26 #35 

What I have done:
Created the following files:
- `requirements.txt`
- `conftest.py`
- `test_extract.py`
- `extract.py`
Written tests for functions in `extract.py`
Written code to pass tests in `extract.py`
Abstracted `apierror` class into separate file

### CODE REVIEW

File: weekly-openmeteo/extract.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import logging
+from requests import get
+from api_error import APIError
+
+URL = "https://api.open-meteo.com/v1/forecast"
+PARAMS = {"latitude": [54.61, 56.49, 52.13, 54.97, 53.78, 53.8,
+                       52.9, 52.48, 52.24, 51.51, 51.28, 50.78],
+	      "longitude": [-6.62, -4.2, -3.78, -1.61, -2.7, -1.54,
+                       -1.23, -1.89, 0.9, -0.13, -0.78, -3.79],
+	      "daily": ["sunrise", "sunset"],"timezone": "Europe/London"}
```
Comment (CalebMathers) [inline weekly-openmeteo/extract.py:11]: This could be for counties rather than regions so we get more precise times for a region as large as Scotland

File: weekly-openmeteo/conftest.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
+'''Contains the configurations for the tests in this folder'''
+import pytest
+
+
+@pytest.fixture
```
Comment (CalebMathers) [inline weekly-openmeteo/conftest.py:5]: Good use of a fixture

File: weekly-openmeteo/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+requests
+psycopg2
+python-dotenv
+pytest
+pylint
+pytest-cov
```
Comment (CalebMathers) [inline weekly-openmeteo/requirements.txt:6]: Good to know the coverage of tests

File: weekly-openmeteo/extract.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+def extract() -> dict:
+    '''Returns the solar data for the given set of coordinates'''
+    response = get(URL, params=PARAMS, timeout=10)
+    logging.info('Request sent.')
+
+    if response.status_code == 200:
+        logging.info('Get request successful.')
```
Comment (CalebMathers) [inline weekly-openmeteo/extract.py:19]: Good use of logging!

File: weekly-openmeteo/extract.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        solar_data = response.json()
+
+        if isinstance(solar_data,list):
+
+            if len(solar_data) == 12 and \
+                all(len(region['daily']['sunrise']) > 0 for region in solar_data) and \
+                all(len(region['daily']['sunset']) > 0 for region in solar_data):
```
Comment (CalebMathers) [inline weekly-openmeteo/extract.py:26]: I like the checks for an existing sunset and sunrise time for each region

**Review by CalebMathers — CHANGES_REQUESTED** [review]

File: weekly-openmeteo/extract.py (hunk 1 above)
Comment (AlexOBarnes) [reply weekly-openmeteo/extract.py]: Good spot I will make these changes

**Review by CalebMathers — APPROVED** [review]

File: weekly-openmeteo/requirements.txt
Comment (olliejth) [inline weekly-openmeteo/requirements.txt:1]: Test coverage is a nice additional feature.

File: weekly-openmeteo/api_error.py
Comment (olliejth) [inline weekly-openmeteo/api_error.py:1]: Interesting use of OOP.

**Review by olliejth — APPROVED** [review]
The API error class is a nice way of handling errors neatly.

File: weekly-openmeteo/api_error.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+'''This module contains the API error class'''
+
+class APIError(Exception):
```
Comment (MurasatoNaoya) [inline weekly-openmeteo/api_error.py:3]: Good use of OOP in a way that doesn't feel forced.

File: weekly-openmeteo/extract.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                 host={ENV["DB_HOST"]} password={ENV["DB_PASSWORD"]} port={ENV["DB_PORT"]}""")
+
+def get_county_coordinates() -> tuple[list[int]]:
+    '''Returns a params dict containing the UK counties coordinates'''
+    query = '''SELECT longitude,latitude FROM county'''
+    with get_connection() as conn:
+        with conn.cursor() as cur:
+            cur.execute(query)
+            counties = cur.fetchall()
+    return [county[0] for county in counties], [county[1] for county in counties]
```
Comment (MurasatoNaoya) [inline weekly-openmeteo/extract.py:23]: Nice list comprehensions for fetching county coordinates.

File: weekly-openmeteo/api_error.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    @message.setter
+    def message(self, new_message):
+        '''Setter for the message property of the APIError class'''
+        self._message = f"{self._message}: {new_message}"
+
+    def define_error(self):
+        '''Returns the associated error message to the status code'''
+        error_messages = {
+            400: "Bad Request: The server could not understand the request.",
```
Comment (MurasatoNaoya) [inline weekly-openmeteo/api_error.py:26]: Extensive list of errors what clear descriptions. Will be useful if we run into issues with extraction later in the project.

File: weekly-openmeteo/extract.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            if len(solar_data) == 106 and \
+                all(len(region['daily']['sunrise']) > 0 for region in solar_data) and \
+                all(len(region['daily']['sunset']) > 0 for region in solar_data):
+                return solar_data
+
+            raise ValueError('Missing Values in returned data')
+
+        raise TypeError('Data returned is of wrong datatype')
+
+    logging.info('Get request unsuccessful.')
```
Comment (MurasatoNaoya) [inline weekly-openmeteo/extract.py:54]: Good use of logging, will help debugging in the terminal if we run into any issues with extraction.

File: weekly-openmeteo/test_extract.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        mock_response._content = json.dumps(valid_response).encode('utf-8')
+        mock_get.return_value = mock_response
+
+        data = request_solar_data(valid_params)
+
+        assert data == valid_response
+
+    @pytest.mark.parametrize("status_code", [400,401,402,403,404,405,408,500])
+    @patch('extract.get')
+    def test_extract_unsuccessful_request(self, mock_get, status_code, valid_params):
```
Comment (MurasatoNaoya) [inline weekly-openmeteo/test_extract.py:27]: I like the use of parametrisation, allowing the testing of a wide variety of error codes.

**Review by MurasatoNaoya — APPROVED** [review]
LGTM.
